doc/dev/tmem.md

```diff
@@ -0,0 +1,1167 @@
+```cpp
+TEST_F(TMemTutorialC, WrongSubpartition2) {
+  NOT_IMPLEMENTED
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigConcreteTensor({32, 2});
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  auto tv2 = set(tv1);
+  auto tv3 = set(tv2);
+  auto tv4 = set(tv3);
+  fusion.addOutput(tv4);
+  tv2->setMemoryType(MemoryType::Tensor);
+
+  tv4->axis(0)->parallelize(ParallelType::TIDx);
+  tv4->axis(1)->parallelize(ParallelType::TIDy);
+  scheduler_utils::parallelizeAllLike(tv4);
+
+  tv2->setTMemDimSepPos(1);
+
+  EXPECT_THAT(
+      [&]() { KernelExecutor().compile(&fusion); },
+      ::testing::ThrowsMessage<std::runtime_error>(::testing::HasSubstr(
+          "Invalid data access pattern in TMem load/store.")));
+} /*
+```
+
+The above example is also invalid because the warp accesses the wrong subpartition.
+In the above example, there are two warps, both accessing subpartition 0, which
+is not allowed.
+
+Now, let's take a look at some valid examples:<!-- */ //-->\
+```cpp
+TEST_F(TMemTutorialR, WarpXYZ) {
+  NOT_IMPLEMENTED
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigConcreteTensor({2, 4, 4, 2});
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  auto tv2 = set(tv1);
+  auto tv3 = set(tv2);
+  auto tv4 = set(tv3);
+  fusion.addOutput(tv4);
+  tv2->setMemoryType(MemoryType::Tensor);
+
+  tv4->axis(0)->parallelize(ParallelType::TIDz);
+  tv4->axis(1)->parallelize(ParallelType::TIDy);
+  tv4->axis(2)->parallelize(ParallelType::TIDx);
+  scheduler_utils::parallelizeAllLike(tv4);
+
+  tv2->setTMemDimSepPos(3);
+
+  if constexpr (verbose) {
+    fusion.printKernel();
+  }
+
+  KernelExecutor ke;
+  ke.compile(&fusion);
+
+  at::Tensor t0 = at::rand({2, 4, 4, 2}, at::kCUDA);
+  auto out = ke.run({t0});
+  EXPECT_TRUE(at::equal(out[0], t0));
+} /*
+```
+
+In the above example, each CTA only has one warp, and this warp is split into
+TIDz, TIDy, and TIDx. The above kernel uses a loop of 2, where each iteration
+accesses a 32x1 box of the tensor memory. This is a valid 32x32b pattern
+.<!-- */ //-->\
+```cpp
+TEST_F(TMemTutorialR, WarpGroupXYZ) {
+  NOT_IMPLEMENTED
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigConcreteTensor({2, 8, 8, 2});
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  auto tv2 = set(tv1);
+  auto tv3 = set(tv2);
+  auto tv4 = set(tv3);
+  fusion.addOutput(tv4);
+  tv2->setMemoryType(MemoryType::Tensor);
+
+  tv4->axis(0)->parallelize(ParallelType::TIDz);
+  tv4->axis(1)->parallelize(ParallelType::TIDy);
+  tv4->axis(2)->parallelize(ParallelType::TIDx);
+  scheduler_utils::parallelizeAllLike(tv4);
+
+  tv2->setTMemDimSepPos(3);
+
+  if constexpr (verbose) {
+    fusion.printKernel();
+  }
+
+  KernelExecutor ke;
+  ke.compile(&fusion);
+
+  at::Tensor t0 = at::rand({2, 8, 8, 2}, at::kCUDA);
+  auto out = ke.run({t0});
+  EXPECT_TRUE(at::equal(out[0], t0));
+} /*
+```
+
+In the above example, each CTA has one warp group. This entire warp group is
```

Comment:
One warp group means a group of consecutive 4 warps?

Reply:
Yes, updated the doc with a description of this.